### PR TITLE
PIM-3625: Added click on notification feature context

### PR DESCRIPTION
**Pull request by ClementGautier**
| Q                 | A
| ----------------- | ---
| Added Specs       | N/A
| Added Behats      | N/A
| Travis CI is ok   | Y
| Changelog updated | N/A

### CODE REVIEW

File: features/Context/AssertionContext.php [hunk 1]
```diff
@@ -604,6 +604,38 @@ public function iShouldHaveNewNotification($count)
...
+     */
+    public function iOpenTheNotificationPanel()
+    {
+        $node = $this->getCurrentPage()->find('css', '#header-notification-widget');
+
+        if ($node->hasClass('open')) {
+            return;
+        }
+
+        $element = $this->getCurrentPage()->find('css', '#header-notification-widget');
```
Comment (phaseinducer) [inline features/Context/AssertionContext.php]: Why finding it again ?

File: features/Context/AssertionContext.php [hunk 2]
```diff
@@ -604,6 +604,38 @@ public function iShouldHaveNewNotification($count)
     }
 
     /**
+     * @When /^I open the notification panel$/
+     */
+    public function iOpenTheNotificationPanel()
+    {
+        $node = $this->getCurrentPage()->find('css', '#header-notification-widget');
```
Comment (sumbobyboys) [inline features/Context/AssertionContext.php]: please spin

File: features/Context/AssertionContext.php [hunk 3]
```diff
@@ -604,6 +604,38 @@ public function iShouldHaveNewNotification($count)
...
+    }
+
+    /**
+     * @When /^I click on the notification "([^"]+)"$/
+     */
+    public function iClickOnTheNotification($message)
+    {
+        $this->iOpenTheNotificationPanel();
+
+        $link = $this->getCurrentPage()->find('css', sprintf('#header-notification-widget .dropdown-menu li>a:contains("%s")', $message));
```
Comment (sumbobyboys) [inline features/Context/AssertionContext.php]: please spin

File: features/Context/AssertionContext.php [hunk 4]
```diff
@@ -604,6 +604,38 @@ public function iShouldHaveNewNotification($count)
...
+
+        if (null === $link) {
+            throw $this->createExpectationException('No notification found on the page.');
```
Comment (phaseinducer) [inline features/Context/AssertionContext.php]: Not really an ExpectationException imho.

File: features/Context/AssertionContext.php (hunk 1 above)
Comment (ClementGautier) [inline features/Context/AssertionContext.php]: bad copy/paste

File: features/Context/AssertionContext.php (hunk 3 above)
Comment (phaseinducer) [inline features/Context/AssertionContext.php]: Maybe this could be refactored (see l. 658)

File: features/Context/AssertionContext.php [hunk 5]
```diff
@@ -604,6 +604,46 @@ public function iShouldHaveNewNotification($count)
...
+
+        if ($node->hasClass('open')) {
+            return;
+        }
+
+        $node->find('css', '.dropdown-toggle')->click();
+
+        // Wait for the footer of the notification panel dropdown to be loaded
+        $this->spin(function() {
+            $footer  = $this->getCurrentPage()->find('css', '#header-notification-widget ul.dropdown-menu p.unspaced');
```
Comment (phaseinducer) [inline features/Context/AssertionContext.php]: "unspaced" is a style class, not a semantic class, so not a good idea to rely on it imo.

**Comment by phaseinducer** [comment]
GTM

File: features/Context/AssertionContext.php [hunk 6]
```diff
@@ -604,6 +604,46 @@ public function iShouldHaveNewNotification($count)
     }
 
     /**
+     * @When /^I open the notification panel$/
+     */
+    public function iOpenTheNotificationPanel()
+    {
+        $node = $this->spin(function () {
```
Comment (sumbobyboys) [inline features/Context/AssertionContext.php]: could you please rename it to $notificationWidget ?

File: features/Context/AssertionContext.php [hunk 7]
```diff
@@ -612,9 +652,11 @@ public function iShouldHaveNewNotification($count)
      */
     public function iShouldSeeNotifications(TableNode $table)
     {
-        $element = $this->getCurrentPage()->find('css', '#header-notification-widget');
-        $element->find('css', '.dropdown-toggle')->click();
-        $this->getMainContext()->wait();
+        $this->iOpenTheNotificationPanel();
+
+        $element = $this->spin(function () {
```
Comment (sumbobyboys) [inline features/Context/AssertionContext.php]: $notificationWidget ?

**Comment by sumbobyboys** [comment]
GTM after comment fixes

**Pull request merged**